lib/truffle/truffle/patching.rb

```diff
@@ -1,34 +1,116 @@
+  PATCHES = { "bundler"                                 => :before,
+              "bundler/version"                         => :after,
+              "bundler/cli/exec"                        => :after,
+              "bundler/compact_index_client/updater"    => :after,
+              "bundler/current_ruby"                    => :after,
+              "bundler/dependency"                      => :after,
+              "bundler/fetcher/downloader"              => :after,
+              "bundler/fetcher/compact_index"           => :after,
+              "bundler/ruby_version"                    => :after,
+              "bundler/source/rubygems"                 => :after,
+
+              "rubygems"                                => :before,
+              "rubygems/ext"                            => :after,
+              "rubygems/package"                        => :after,
+              "rubygems/remote_fetcher"                 => :after,
+              "rubygems/request"                        => :after,
+              "rubygems/request_set/gem_dependency_api" => :after,
+
+              "rspec/support"                           => :after,
+
+              "unf/normalizer_cruby"                    => :before,
+  }
+
+  PATCHES.each do |file, _|
+    unless File.exist? File.join(DIR, file) + '.rb'
+      raise "file #{file} is missing in #{DIR}"
+    end
+  end
+
+  Dir["#{DIR}/**/*.rb"].each do |file|
+    relative_path = file[(DIR.size + 1)..-4]
+    unless PATCHES.key? relative_path
+      raise "file #{relative_path} is not declared in PATCHES"
+    end
+  end
+
+  def self.find_absolute_path(relative_path)
+    $LOAD_PATH.each do |base|
+      candidate = File.join(base, relative_path) + '.rb'
+      return candidate if File.exist? candidate
+    end
+    raise "absolute path for #{relative_path} was not found in #{$LOAD_PATH}"
+  end
+
+  def self.find_relative_path(absolute_path)
+    return nil unless File.exist? absolute_path
+
+    $LOAD_PATH.each do |base|
+      if absolute_path.start_with?(base)
+        return File.basename absolute_path[(base.size)..-1], '.*'
+      end
+    end
+
+    nil
+  end
+
+  def self.required?(relative_path)
+    $LOADED_FEATURES.include? find_absolute_path relative_path
+  end
 end
 
 module Kernel
 
   private
 
-  alias_method :require_without_truffle_patching, :require
+  alias_method :require_without_truffle_patching, :gem_original_require
 
-  def require(path)
-    required = begin
-      require_without_truffle_patching path
-    rescue LoadError
-      if Truffle::Patching::TRUFFLE_PATCHES[path]
-        true # Pretend the file required so we attempt patching with something that should avoid the LoadError (e.g., replacing code using C extensions).
+  def gem_original_require(path)
+    path           = Rubinius::Type.coerce_to path, String, :to_str
+    extname        = File.extname path
+    path_no_suffix = path[0..(-1-extname.size)]
+    relative_path  = if path_no_suffix.start_with?('/')
```

Comment:
I expect it will take a long time before we can remove them all.
I don't think it's good to add extra non-standard require before user code.
For instance, `require "pathname"` will return false for us, true for MRI if in user code.
It also affects startup.
But it is very tempting and convenient, I give you that.